kernel/pipe.c

```diff
@@ -0,0 +1,189 @@
+		return -EAGAIN;
+	}
+	pipe->waiting++;
+	z_pend_curr(&pipe->lock, *key, waitq, timeout);
+	*key = k_spin_lock(&pipe->lock);
+	pipe->waiting--;
+	if (unlikely(!(pipe->flags & PIPE_FLAG_OPEN))) {
+		return -EPIPE;
+	} else if (unlikely(pipe->flags & PIPE_FLAG_RESET)) {
+		if (!pipe->waiting) {
+			pipe->flags &= ~PIPE_FLAG_RESET;
+		}
+		return -ECANCELED;
+	} else if (!cond(pipe)) {
+		return 0;
+	}
+	return -EAGAIN;
+}
+
+static void notify_waiter(_wait_q_t *waitq)
+{
+	struct k_thread *thread_to_unblock = z_unpend_first_thread(waitq);
+
+	if (likely(thread_to_unblock)) {
+		z_ready_thread(thread_to_unblock);
+	}
+}
+
+void z_impl_k_pipe_init(struct k_pipe *pipe, uint8_t *buffer, size_t buffer_size)
+{
+	ring_buf_init(&pipe->buf, buffer_size, buffer);
+	pipe->flags = PIPE_FLAG_OPEN;
+
+	pipe->lock = (struct k_spinlock){0};
+	z_waitq_init(&pipe->data);
+	z_waitq_init(&pipe->space);
+}
+
+int z_impl_k_pipe_write(struct k_pipe *pipe, const uint8_t *data, size_t len, k_timeout_t timeout)
+{
+	int rc;
+	k_spinlock_key_t key;
+
+	__ASSERT_NO_MSG(pipe != NULL);
+	__ASSERT_NO_MSG(data != NULL);
+
+	key = k_spin_lock(&pipe->lock);
+	if (unlikely(pipe_full(pipe))) {
+		rc = wait_for(&pipe->space, pipe, &key, pipe_full, timeout);
+		if (rc) {
+			k_spin_unlock(&pipe->lock, key);
+			return rc;
+		}
+	}
+
+	if (!(pipe->flags & PIPE_FLAG_OPEN)) {
+		k_spin_unlock(&pipe->lock, key);
+		return -EPIPE;
+	}
+
+	rc = ring_buf_put(&pipe->buf, data, len);
+	if (rc) {
+		notify_waiter(&pipe->data);
+	}
+	k_spin_unlock(&pipe->lock, key);
+	return rc;
+}
+
+int z_impl_k_pipe_read(struct k_pipe *pipe, uint8_t *data, size_t len, k_timeout_t timeout)
+{
+	int rc;
+	k_spinlock_key_t key;
+
+	__ASSERT_NO_MSG(pipe != NULL);
+	__ASSERT_NO_MSG(data != NULL);
+
+	key = k_spin_lock(&pipe->lock);
+	if (pipe_empty(pipe) && pipe->flags & PIPE_FLAG_OPEN) {
+		rc = wait_for(&pipe->data, pipe, &key, pipe_empty, timeout);
+		if (rc && rc != -EPIPE) {
+			k_spin_unlock(&pipe->lock, key);
+			return rc;
+		}
+	}
+	if (pipe_empty(pipe) && !(pipe->flags & PIPE_FLAG_OPEN)) {
+		k_spin_unlock(&pipe->lock, key);
+		return -EPIPE;
+	}
+
+	rc = ring_buf_get(&pipe->buf, data, len);
+	if (rc) {
+		notify_waiter(&pipe->space);
+	}
+	k_spin_unlock(&pipe->lock, key);
+	return rc;
+}
+
+void z_impl_k_pipe_flush(struct k_pipe *pipe)
+{
+	__ASSERT_NO_MSG(pipe != NULL);
+	K_SPINLOCK(&pipe->lock) {
+		pipe->flags |= PIPE_FLAG_RESET;
+		ring_buf_reset(&pipe->buf);
+		z_unpend_all(&pipe->data);
+		z_unpend_all(&pipe->space);
+	}
+}
+
+void z_impl_k_pipe_close(struct k_pipe *pipe)
+{
+	__ASSERT_NO_MSG(pipe != NULL);
+	K_SPINLOCK(&pipe->lock) {
+		pipe->flags = 0;
+		z_unpend_all(&pipe->data);
+		z_unpend_all(&pipe->space);
+	}
+}
+
+#ifdef CONFIG_USERSPACE
+void z_vrfy_k_pipe_init(struct k_pipe *pipe, uint8_t *buffer, size_t buffer_size)
+{
+	K_OOPS(K_SYSCALL_OBJ(pipe, K_OBJ_PIPE));
+	K_OOPS(K_SYSCALL_MEMORY_WRITE(buffer, buffer_size));
+	K_OOPS(K_SYSCALL_MEMORY_READ(buffer, buffer_size));
```

Comment:
You don't need both of these.  K_SYSCALL_MEMORY_WRITE checks for read permissions too.